src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,569 @@
+    end
+  end
+
+  # Represents the lock-free ring queue for consuming responses from the Kernel. It lives in shared memory.
+  #
+  # The queue has an `@array` of size `@size`. The size is a power of two to make calculations easier.
+  # `@mask` is just `@size - 1`, useful to bring a index into the array range.
+  #
+  # We keep 2 pointers into the queue:
+  # - `@tail`: The next item the Kernel will write.
+  # - `@head`: The next item we will read, unless it is equal to `@tail`.
+  # Both pointers only move forward increasing by one each time (except for overflow). `@mask` is
+  # used on every access to `@array`.
+  #
+  # `@head` and `@tail` are pointers to where these values are actually stored in shared memory. Only the
+  # Kernel updates ´@tail.value`, so we must always use ACQUIRE atomics to read from it. Likewise only
+  # we update `@head.value`, so updating it requires RELEASE atomics.
+  #
+  # `@overflow.value` stores the number of responses the Kernel discarted because there wasn't room in the
+  # queue to insert them. Since Linux 5.4 it should be always zero (NODROP feature).
+  private struct CompletionQueue
+    @mask : UInt32
+    @size : UInt32
+
+    def initialize(@fd : Int32, base : Void*, offsets : Syscall::IoCqringOffsets)
+      @head = Pointer(UInt32).new(base.address + offsets.head)
+      @tail = Pointer(UInt32).new(base.address + offsets.tail)
+      @mask = Pointer(UInt32).new(base.address + offsets.ring_mask).value
+      @size = Pointer(UInt32).new(base.address + offsets.ring_entries).value
+      @array = Slice(Syscall::IoUringCqe).new(Pointer(Syscall::IoUringCqe).new(base.address + offsets.cqes), @size)
+      @overflow = Pointer(UInt32).new(base.address + offsets.overflow)
+
+      if @head.value != 0 || @tail.value != 0
+        raise RuntimeError.new("CompletionQueue should start empty")
+      end
+
+      if @size != @mask + 1
+        raise RuntimeError.new("CompletionQueue size and mask are mismatched")
+      end
+    end
+
+    # TODO: Remove it.
+    def inspect(io)
+      tail = Atomic::Ops.load(@tail, :acquire, false)
+      io << "CompletionQueue(H: #{@head.value}, T: #{tail}, #{size}/#{@size})"
+    end
+
+    # Optimistic view of how many itens are queued. As soon as this is computed the Kernel might insert
+    # new items from one of its threads.
+    def size
+      Atomic::Ops.load(@tail, :acquire, false) - @head.value
+    end
+
+    # Returns the number of responses the Kernel discarted because there wasn't room in the
+    # queue to insert them. Since Linux 5.4 it should be always zero (NODROP feature). If this is
+    # not zero, then there is no way to know which event was dropped and we should abort().
+    def overflow
+      Atomic::Ops.load(@overflow, :acquire, false)
+    end
+
+    # Consume and yield each currently available item of the queue, leaving room for the kernel to write.
+    def consume_all
+      tail = Atomic::Ops.load(@tail, :acquire, false)
```

Comment:
liburing does that as an implementation detail to skip any cqe with `LIBURING_UDATA_TIMEOUT` user data. This is because `io_uring_wait_cqes` was a timeout argument that works by sending a timeout sqe that should never be visible to the application.

> Like io_uring_wait_cqe(), except it accepts a timeout value as well. Note
> that an sqe is used internally to handle the timeout. For kernel doesn't
> support IORING_FEAT_EXT_ARG, applications using this function must never
> set sqe-&gt;user_data to LIBURING_UDATA_TIMEOUT!

See https://github.com/axboe/liburing/blob/master/src/queue.c#L298.

We don't need this because we never skip any cqe, all of them are relevant.